programs/mango-v4/src/instructions/account_buyback_fees_with_mngo.rs

```diff
@@ -0,0 +1,158 @@
+use anchor_lang::prelude::*;
+use checked_math as cm;
+use fixed::types::I80F48;
+
+use crate::accounts_zerocopy::*;
+use crate::error::MangoError;
+use crate::state::*;
+
+use crate::accounts_ix::*;
+
+pub fn account_buyback_fees_with_mngo(
+    ctx: Context<AccountBuybackFeesWithMngo>,
+    max_buyback: u64,
+) -> Result<()> {
+    // Cannot buyback from yourself
+    require_keys_neq!(
+        ctx.accounts.account.key(),
+        ctx.accounts.dao_account.key(),
+        MangoError::SomeError
+    );
+
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    // account constraint #1
+    require!(
+        account.fixed.is_owner_or_delegate(ctx.accounts.owner.key()),
+        MangoError::SomeError
+    );
+
+    let mut dao_account = ctx.accounts.dao_account.load_full_mut()?;
+
+    let group = ctx.accounts.group.load()?;
+
+    let mut mngo_bank = ctx.accounts.mngo_bank.load_mut()?;
+    let mut fees_bank = ctx.accounts.fees_bank.load_mut()?;
+
+    let bonus_factor = I80F48::from_num(group.buyback_fees_mngo_bonus_factor);
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+    // quick return if nothing to buyback
+    let mut max_buyback = {
+        let dao_fees_token_position = dao_account.ensure_token_position(fees_bank.token_index)?.0;
+        let dao_fees_native = dao_fees_token_position.native(&fees_bank);
+        I80F48::from_num::<u64>(max_buyback.min(account.fixed.buyback_fees_accrued))
+            .min(dao_fees_native)
```

Comment:
I'm worried about what happens here if dao_fees_native is negative. We need to either require it's not, or change the check below to max_buyback <= 0 (the latter would be more defensive anyway)